### PR TITLE
code review

### PR DESCRIPTION
**Pull request by andydam**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,9 +35,11 @@
   "dependencies": {
     "body-parser": "^1.18.2",
```
Comment (cjules0609) [inline package.json:36]: You don't user body-parser anymore right.

File: products/analytics/details.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+const { insert } = require('./insert');
+
+const view = (userId, productId, time) =>
+  insert(userId, 'VIEW', { PRODUCT: productId }, time, 10000);
+
```
Comment (brycessorensen) [inline products/analytics/details.js:5]: Please add error handling

File: products/analytics/reviews.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+const { insert } = require('./insert');
+
+const post = (userId, productId, time) => insert(userId, 'REVIEW', { PRODUCT: productId }, time);
```
Comment (brycessorensen) [inline products/analytics/reviews.js:3]: Please add error handling

File: products/cache.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+const util = require('util');
+
+const client = redis.createClient(process.env.REDIS_URL || 'redis://127.0.0.1:6379');
+client.on('error', console.error);
+
+client.getAsync = util.promisify(client.get);
+client.setAsync = util.promisify(client.set);
+
+const memoize = async (func, ...args) => {
+  const key = func.name + JSON.stringify(args);
```
Comment (brycessorensen) [inline products/cache.js:11]: Can use template literal instead of string concatenation

File: products/database/reviews.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  );
+
+const getByUserId = userId =>
+  connection.queryAsync(
+    'SELECT text, created_at, stars, credibility, is_customer, product_id FROM product_reviews WHERE user_id = ? ORDER BY created_at DESC',
+    [userId],
+  );
+
+const insert = (userId, productId, text, createdAt, stars, isCustomer, isProfessional) =>
+  connection
```
Comment (cjules0609) [inline products/database/reviews.js:16]: Last two connection.queryAsync inline, this one has \n

File: products/database/analytics.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+const connection = require('../../database');
+
+const insert = (userId, activity, details, time, duration) =>
+  connection
+    .queryAsync(
+      'INSERT INTO product_analytics (user_id, activity, details, time, duration) VALUES (?, ?, ?, ?, ?)',
+      [userId, activity, details, time, duration],
+    )
+    .then(results => results.insertId);
```
Comment (brycessorensen) [inline products/database/analytics.js:9]: Can use async await instead of .then

File: products/database/reviews.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    .queryAsync(
+      'INSERT INTO product_reviews (user_id, product_id, text, created_at, stars, is_customer, is_professional) VALUES (?, ?, ?, ?, ?, ?, ?)',
+      [userId, productId, text, createdAt, stars, isCustomer, isProfessional],
+    )
+    .then(results => results.insertId);
+
+module.exports = { getByProductId, getByUserId, insert };
```
Comment (brycessorensen) [inline products/database/reviews.js:23]: Inconsistent module exports styling. This example is inline, whereas others are not. Please make consistent

File: products/format.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  description: product._source.description,
+  categories: product._source.categories,
+});
+
+const error = error => ({
+  message: error.message,
+  statusCode: error.status,
+});
+
+module.exports = { product, error };
```
Comment (brycessorensen) [inline products/format.js:15]: Please match exports styling to be similar to all files

File: products/index.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    return res.status(500).json(err);
+  }
+});
+
+router.get('/products/related/:id', async (req, res) => {
+  if (!collaborativePath && !contentPath) {
+    return res.sendStatus(504);
+  }
+  try {
+    const urls = [`${contentPath}${req.params.id}`, `${contentPath}${req.params.id}`];
```
Comment (brycessorensen) [inline products/index.js:108]: Use object destructuring to pull id out of req.params

**Review by cjules0609 — COMMENTED** [review]
Clean, concise and modular.

File: products/index.js [hunk 10]
```diff
@@ -1,6 +1,10 @@
 const express = require('express');
+const fetch = require('node-fetch');
 const jsonParser = require('body-parser').json();
 
+const collaborativePath = process.env.COLLAB_PATH;
```
Comment (emmayunliu) [inline products/index.js]: collaborative filtering by user_id is also available. endpoint is collaborative/user.

File: products/database/reviews.js [hunk 11]
```diff
@@ -2,13 +2,13 @@ const connection = require('../../database');
 
 const getByProductId = productId =>
   connection.queryAsync(
-    'SELECT text, user_id, created_at, stars, credibility, is_customer FROM product_reviews WHERE product_id = ?',
+    'SELECT text, user_id, created_at, stars, credibility, is_customer FROM product_reviews WHERE product_id = ? ORDER BY credibility DESC, created_at DESC',
```
Comment (emmayunliu) [inline products/database/reviews.js]: could add distinct before text in case that duplicated reviews exist

**Pull request closed without merging**